### ISSUE TITLE
Feature: Add switch to Invoke-Pester to automatically re-run failed tests

### ISSUE THREAD

**Issue by SteveL-MSFT** [pester/Pester#1161]
## 1. General summary of the issue

To help identify flaky tests (tests that fail sometimes inconsistently rather than failing consistently), it would help to have Pester have opt-in ability to rerun failed tests (as it's being executed).  Need a way to indicate the test failed or passed on the rerun so that the test gets fixed to be predictable and consistent.

## 2. Describe Your Environment

Pester version     : 4.4.0 /Users/steve/.local/share/powershell/Modules/Pester/4.4.0/Pester.psd1
PowerShell version : 6.2.0-preview.2
OS version         : Unix 18.2.0.0

## 5. Possible Solution

Perhaps add a `-RerunFailedTests` parameter to `Invoke-Pester` that accepts a `uint` on number of times to rerun.  Certainly this requires test cases to be idempotent which is a best practice anyways.  Currently, one would have to rerun the entire test script which is not optimal if only 1 test case out of many fails on occasion.

**Comment by bergmeister** [comment]
@SteveL-MSFT  The `VsTestV2` task in Azure DevOps already provides that logic, therefore it might be better to rather try to integrate Pester with `vstest.console.exe` using a so called test adapter (this is what e.g. `NUnit` does to make their test framework usable from VisualStudio/vstest.console.exe . WDYT?

**Comment by SteveL-MSFT** [comment]
@bergmeister We've been in discussions with AzDevOps team about this topic which is where this came up actually.  They just need some ability to re-run tests (preferably without re-running the whole thing) and AzDevOps will handle all the reporting aspects.  Currently, they already have reporting on test case pass/fail history so they can identify potentially unreliable tests, but this is over a period of time running the complete test suite.  Running only the failed tests would be a perf optimization.

Integration with `vstest.console.exe` seems like a nice to have and should be a separate thing rather than requiring it.

**Comment by michalporeba** [comment]
That ability to rerun a specific test, (and get perhaps some metadata / associated code of a failed test) should be very helpful to all sorts of extensions and 'wrappers' around Pester.

**Comment by nohwnd** [comment]
v5-alpha1 actually can already do it (in a bit crude way).  You can see me testing [re-running failed tests here](https://github.com/pester/Pester/blob/4748eaabd2d035c403f54e0e9267fa024139575e/new-runtimepoc/Pester.Runtime.Tests.ps1#L430). I agree that it opens up a lot of possibilities 🙂 

At the moment those published functions do not take the result object + paths (to keep the api smaller and simpler), instead test discovery is done every time from scratch. 

But internally the limitation is not there, if I add functions that clean a dirty discover/run result, it could easily take a cached discovery object and run with different filters.  Which is useful not only for this scenario, but for example if you want to observe filesystem, discover tests in a single file, join that discovery result to your cached results and re-run all previously failed tests in all files. Or when you have a huge test suite where discovery takes signifcant amount of time, and multiple configurations to run it with.

**Comment by nohwnd** [comment]
Yes, that is what my comment says I believe.

**Comment by ArmaanMcleod** [comment]
@nohwnd @fflaten I can probably help with this. I've actually needed this functionality to detect flaky tests for reporting purposes. Being able to rerun failed tests would be super useful. 

I like the idea of using a cached discovery object from `-PassThru` and re-running failed(possibly even skipped) tests via a `-ReRun` filter, similar to the approach described in the other closed issue. 

Just wondering how we would include this new option into `Invoke-Pester`? I'm assuming this would need to be available in the `Advanced` parameter set, not sure about `Simple`. I can see this needing two parameters at least, a `Pester.Run` object of the previous run and a filter to determine what to rerun(Failed, Skipped etc.). 

Or would it be better to publish a new cmdlet for this?

**Comment by nohwnd** [comment]
@ArmaanMcleod which case are you after? Re-running failed tests can be done by taking the result object, and running only the failed tests by line filter. We don't have a function to merge-in the failed tests results after re-run, so you only get a single result object.

There are some limitations that I did not realize when writing the comment about the cached discovery object, all of them connected to generated tests that depend on dynamic data.

**Comment by ArmaanMcleod** [comment]
@nohwnd Yeah your right. I didn't see that `StartLine` is included in the result object for failed tests 😄 . I should have looked closer. 

Something like below works fine for my usecase:

```pwsh
$testPath = "C:\sample.Tests.ps1"

Invoke-Pester -Path $testPath -PassThru

$failedTestLines = $result.Failed | 
    Select-Object -Property StartLine -Unique | 
        ForEach-Object { "${testPath}:$_" }

$pesterConfig = New-PesterConfiguration -Hashtable @{
    Run = @{
        Container = New-PesterContainer -Path $testPath
    }
    Filter = @{
        Line = $failedTestLines
    }
}

Invoke-Pester -Configuration $pesterConfig
```

I guess this could be wrapped into another option, but might be unnecessary since the above is easy enough to do if you are familiar with the line filter.

**Comment by nohwnd** [comment]
Do you actually get a failure in the container? Failed container is populated only when there is a critical failure **directly in** a container, not when there are any failed tests.
